dm/worker/subtask.go

```diff
@@ -247,70 +250,70 @@ func (st *SubTask) callCurrCancel() {
+	errs := make([]*pb.ProcessError, 0, 2)
+	for _, err := range result.Errors {
+		if !unit.IsCtxCanceledProcessErr(err) {
+			errs = append(errs, err)
 		}
-		result.Errors = errs
+	}
+	result.Errors = errs
 
-		st.callCurrCancel() // dm-unit finished, canceled or error occurred, always cancel processing
+	st.callCurrCancel() // dm-unit finished, canceled or error occurred, always cancel processing
 
-		if len(result.Errors) == 0 && st.Stage() == pb.Stage_Pausing {
-			st.setResult(&result) // save result
-			return                // paused by external request
+	var (
+		cu    = st.CurrUnit()
+		stage pb.Stage
+	)
+
+	// update the stage according to result
+	if len(result.Errors) == 0 {
+		switch st.Stage() {
+		case pb.Stage_Pausing:
+			// paused by st.Pause
+			stage = pb.Stage_Paused
+		case pb.Stage_Stopping:
+			// stopped by st.Close
+			stage = pb.Stage_Stopped
+		default:
+			// process finished with no error
+			stage = pb.Stage_Finished
 		}
+	} else {
+		// error occurred, paused
+		stage = pb.Stage_Paused
+	}
+	st.setStageAndResult(stage, &result)
 
-		var (
-			cu    = st.CurrUnit()
-			stage pb.Stage
-		)
-		if len(result.Errors) == 0 {
-			if result.IsCanceled {
-				stage = pb.Stage_Stopped // canceled by user
-			} else {
-				stage = pb.Stage_Finished // process finished with no error
-			}
+	st.l.Info("unit process returned", zap.Stringer("unit", cu.Type()), zap.Stringer("stage", stage), zap.String("status", st.StatusJSON()))
+
+	switch stage {
+	case pb.Stage_Finished:
+		cu.Close()
+		nu := st.getNextUnit()
+		if nu == nil {
+			// Now, when finished, it only stops the process
+			// if needed, we can refine to Close it
+			st.l.Info("all process units finished")
 		} else {
-			stage = pb.Stage_Paused // error occurred, paused
+			st.l.Info("switching to next unit", zap.Stringer("unit", cu.Type()))
+			st.setCurrUnit(nu)
+			// NOTE: maybe need a Lock mechanism for sharding scenario
+			st.run() // re-run for next process unit
 		}
-		st.setStageAndResult(stage, &result)
-
-		st.l.Info("unit process returned", zap.Stringer("unit", cu.Type()), zap.Stringer("stage", stage), zap.String("status", st.StatusJSON()))
-
-		switch stage {
-		case pb.Stage_Finished:
-			cu.Close()
-			nu := st.getNextUnit()
-			if nu == nil {
-				// Now, when finished, it only stops the process
-				// if needed, we can refine to Close it
-				st.l.Info("all process units finished")
-			} else {
-				st.l.Info("switching to next unit", zap.Stringer("unit", cu.Type()))
-				st.setCurrUnit(nu)
-				// NOTE: maybe need a Lock mechanism for sharding scenario
-				st.run() // re-run for next process unit
-			}
-		case pb.Stage_Stopped:
-		case pb.Stage_Paused:
-			for _, err := range result.Errors {
-				st.l.Error("unit process error", zap.Stringer("unit", cu.Type()), zap.Reflect("error information", err))
-			}
+	case pb.Stage_Stopped:
+		// the caller will close current unit and more units after it, so we don't call cu.Close here.
```

Comment:
In `subtask.Close` we call `st.closeUnits` but in `subtask.Pasuse` we don't call `cu.Pause()`

Perhaps the logic of subtask close, pause could be unified? put both operations in the `setStageAndResult` function？

Reply:
StageAndResult is for subtask, Close is for unit.

Reply:
`subtask.Close` didn't use `setStageAndResult`